### ISSUE TITLE
Exponential window size & memory allocation growth + crash when scaling window on Retina OS X

### ISSUE THREAD

**Issue by pushfoo** [kivy/kivy#6708]
### Versions

* Python:  3.7.5
* OS: 10.12.6
* Kivy: 1.11.1
* Kivy installation method: pip

### Description

**What are you trying to get done?**
Lock resizing of a window to an aspect ratio on an OS X desktop 

**What has happened? / What went wrong?**
Resizing causes:
1. a hang
2. exceptions mentioning failed attachment and very large & negative texture sizes
3. (usually) a crash.

Items are also placed incorrectly on resize.

Listening in on resize events shows exponential growth of the window size with each repeated resize event fired when dragging the window edge:

```
Window size set to: (1600, 1200)
Window size set to: (3186, 1778)
Window size set to: (6372, 3584)
Window size set to: (12744, 7168)
Window size set to: (25488, 14336)
```

**What stops it**
Dividing window size by two before setting it. My suspicion is that this is a runaway exponentiation version of the same issue as faced [here](https://github.com/ElliotGarbus/KivyWindowSize):

> When saving the window size to the .ini file, the Window.size must be divided by Metrics.density. If this not done, and Metrics.density == 2, for example, the window size will double each time the program is opened.

Given that element placement is also affected, a number of other OS X specific issues may be related:
- #6144 
- #6446 

There is also a [closed issue](https://github.com/kivy/kivy/issues/6407) by the author of the quote above that _may_ be a reference to this issue, but it's unclear whether the fix was merged for whatever problem was encountered.

### Code and Logs
Switchable crasher:
```import kivy
from kivy import Config

from kivy.app import App
from kivy.core.window import Window
from kivy.metrics import Metrics
from kivy.uix.screenmanager import Screen, ScreenManager


CRASH = True
ASPECT_RATIO = 0.5625

kivy.require('1.11.0')

class AspectRatioScreen(Screen):
    def on_size(self, a, b):
        """
        Attempts to keep an aspect ratio
        """
        current_ratio = Window.height / Window.width
        if current_ratio != ASPECT_RATIO and Window.fullscreen is False:

            if CRASH: # unmodified size causes the issue
                Window.size = Window.width, Window.width * ASPECT_RATIO

            else: # dividing size by two or Metrics.density causes everything to work
                Window.size = \
                    Window.width / Metrics.density,\
                    (Window.width / Metrics.density) * ASPECT_RATIO


class Crash(App):
    def build(self):
        w = ScreenManager()
        Config.set('graphics', 'height', '200')
        Config.set('graphics', 'height', '200')

        w.add_widget(AspectRatioScreen(name="main"))
        return w

if __name__ == "__main__":
    Crash().run()
```
Example errors:
```
 Exception: FBO Initialization failed: Incomplete attachment (36054)
 Exception ignored in: 'kivy.graphics.fbo.Fbo.raise_exception'
 Exception: FBO Initialization failed: Incomplete attachment (36054)
 Exception: FBO Initialization failed: Incomplete attachment (36054)
 Exception ignored in: 'kivy.graphics.fbo.Fbo.raise_exception'
 Exception: FBO Initialization failed: Incomplete attachment (36054)
 Exception: FBO Initialization failed: Incomplete attachment (36054)
 Exception ignored in: 'kivy.graphics.fbo.Fbo.raise_exception'
 Exception: FBO Initialization failed: Incomplete attachment (36054)
 Exception: FBO Initialization failed: Incomplete attachment (36054)
 Exception ignored in: 'kivy.graphics.fbo.Fbo.raise_exception'
 Exception: FBO Initialization failed: Incomplete attachment (36054)
 Exception: FBO Initialization failed: Incomplete attachment (36054)
 Exception ignored in: 'kivy.graphics.fbo.Fbo.raise_exception'
 Exception: FBO Initialization failed: Incomplete attachment (36054)
 Exception: FBO Initialization failed: Incomplete attachment (36054)
 Exception ignored in: 'kivy.graphics.fbo.Fbo.raise_exception'
 Exception: FBO Initialization failed: Incomplete attachment (36054)
python(75016,0x7fffbc3063c0) malloc: *** mach_vm_map(size=18446744073691299840) failed (error code=3)
*** error: can't allocate region
*** set a breakpoint in malloc_error_break to debug
 Exception: Unable to allocate memory for texture (size is -18251776)
```

**Comment by tito** [comment]
Can you try using Window.system_size (list of system width/height) to see if you replicate it too?

**Comment by pushfoo** [comment]
I'm not entirely sure what you mean. I've guessed a possibilities and tried them.
I tried creating an on_system_size handler method, but it doesn't seem to ever get called in response to resize events.
If you meant scaling via the system_size property, the following code works without applying a division:
```python
import kivy
from kivy import Config

from kivy.app import App
from kivy.core.window import Window
from kivy.metrics import Metrics
from kivy.uix.screenmanager import Screen, ScreenManager


ASPECT_RATIO = 0.5625

kivy.require('1.11.0')

class AspectRatioScreen(Screen):

    def on_system_size(self, a, b):
        #never actually gets called
        print(f"system_size set via system_size property: {Window.system_size}")

    def on_size(self, screen, size_tuple):
        """
        Attempts to keep an aspect ratio
        """
        print(f"System size at start of on_size handler: {Window.system_size}")
        w, h = Window.system_size
        ratio = h / w
        #enforce aspect ratio
        if ratio != ASPECT_RATIO and Window.fullscreen is False:
            Window.system_size = w, w * ASPECT_RATIO

        print(f"System size at end of on_size handler: {Window.system_size}")

class NoCrash(App):
    def build(self):
        w = ScreenManager()
        Config.set('graphics', 'width', '200')
        Config.set('graphics', 'height', '200')

        w.add_widget(AspectRatioScreen(name="main"))
        return w

if __name__ == "__main__":
    NoCrash().run()
```
If you meant something else, let me know and i'll gladly try it.

**Comment by Zen-CODE** [comment]
@tito Bump. Any comment on the above?

**Comment by misl6** [comment]
This should have been fixed via #8630 , however as @tito suggested, even if now `Window.size` works as expected (but consider potential issues with rounding), the suggested way to set the `Window` size is by using `Window.system_size`.

Please also take a look at the refreshed docs regarding these properties:
- https://kivy.org/doc/master/api-kivy.core.window.html#kivy.core.window.WindowBase.system_size
- https://kivy.org/doc/master/api-kivy.core.window.html#kivy.core.window.WindowBase.size.

Can you please confirm that the issue does not happen anymore on your side with the latest `master`?
